ksqldb-engine/src/main/java/io/confluent/ksql/util/PersistentQueriesInSharedRuntimesImpl.java

```diff
@@ -0,0 +1,320 @@
+      final Set<SourceName> sourceNames,
+      final String executionPlan,
+      final String queryApplicationId,
+      final Topology topology,
+      final SharedKafkaStreamsRuntime sharedKafkaStreamsRuntime,
+      final QuerySchemas schemas,
+      final Map<String, Object> overriddenProperties,
+      final QueryId queryId,
+      final Optional<MaterializationProviderBuilderFactory.MaterializationProviderBuilder>
+          materializationProviderBuilder,
+      final ExecutionStep<?> physicalPlan,
+      final ProcessingLogger processingLogger,
+      final DataSource sinkDataSource,
+      final Listener listener) {
+    // CHECKSTYLE_RULES.ON: ParameterNumberCheck
+    this.statementString = Objects.requireNonNull(statementString, "statementString");
+    this.executionPlan = Objects.requireNonNull(executionPlan, "executionPlan");
+    this.queryApplicationId = Objects.requireNonNull(queryApplicationId, "queryApplicationId");
+    this.topology = Objects.requireNonNull(topology, "kafkaTopicClient");
+    this.sharedKafkaStreamsRuntime =
+        Objects.requireNonNull(sharedKafkaStreamsRuntime, "kafkaStreamsBuilder");
+    this.sinkDataSource = requireNonNull(sinkDataSource, "sinkDataSource");
+    this.schemas = requireNonNull(schemas, "schemas");
+    this.overriddenProperties =
+        ImmutableMap.copyOf(
+            Objects.requireNonNull(overriddenProperties, "overriddenProperties"));
+    this.sourceNames = Objects.requireNonNull(sourceNames, "sourceNames");
+    this.queryId = Objects.requireNonNull(queryId, "queryId");
+    this.processingLogger = requireNonNull(processingLogger, "processingLogger");
+    this.physicalPlan = requireNonNull(physicalPlan, "physicalPlan");
+    this.resultSchema = requireNonNull(schema, "schema");
+    this.materializationProviderBuilder =
+        requireNonNull(materializationProviderBuilder, "materializationProviderBuilder");
+    this.listener = requireNonNull(listener, "listen");
+  }
+
+  @Override
+  public DataSource.DataSourceType getDataSourceType() {
+    return sinkDataSource.getDataSourceType();
+  }
+
+  @Override
+  public KsqlTopic getResultTopic() {
+    return sinkDataSource.getKsqlTopic();
+  }
+
+  @Override
+  public SourceName getSinkName() {
+    return sinkDataSource.getName();
+  }
+
+  @Override
+  public QuerySchemas getQuerySchemas() {
+    return schemas;
+  }
+
+  @Override
+  public PhysicalSchema getPhysicalSchema() {
+    return resultSchema;
+  }
+
+  @Override
+  public ExecutionStep<?> getPhysicalPlan() {
+    return physicalPlan;
+  }
+
+  @Override
+  public DataSource getSink() {
+    return sinkDataSource;
+  }
+
+  @Override
+  public ProcessingLogger getProcessingLogger() {
+    return processingLogger;
+  }
+
+  @Override
+  public Optional<Materialization> getMaterialization(
+      final QueryId queryId,
+      final QueryContext.Stacker contextStacker) {
+    return null;
+  }
+
+  @Override
+  public void restart() {
```

Comment:
Just FYI, in https://github.com/confluentinc/ksql/pull/7627/files we removed the `restart()` func -- I think this is not required for binpacking either but in case it does, @wcarlson5 please raise it to the other side for attention.

Reply:
I don't think its required, but it is the only way to get a fresh streams object with bin-packing so it might be nice to have. I am fine either way but @rodesai might have a perfernce.